Update internal code for torch.linalg.solve

IvanYashchuk opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#56613 Update internal code for torch.linalg.solve**

Replace linalg_solve_helper with `lu_stub` + `lu_solve_stub`.
Once `lu_stub` and `lu_solve_stub` have cuSOLVER-based codepath,
`torch.linalg.solve` will have it as well.

Differential Revision: [D28627408](https://our.internmc.facebook.com/intern/diff/D28627408)

IvanYashchuk committed:
Update internal code for torch.linalg.solve

Replace linalg_solve_helper with `lu_stub` + `lu_solve_stub`.
Once `lu_stub` and `lu_solve_stub` have cuSOLVER-based codepath,
`torch.linalg.solve` will have it as well.

[ghstack-poisoned]

IvanYashchuk commented on test/test_linalg.py:
```diff
@@ -3255,6 +3255,7 @@ def solve_test_helper(self, A_dims, b_dims, device, dtype):
         A = random_fullrank_matrix_distinct_singular_value(*A_dims, dtype=dtype).to(device)
         return b, A
 
+    @onlyOnCPUAndCUDA
```
Skips meta device because of `RuntimeError: DispatchStub: unsupported device typemeta`.

walterddr replied:
Seems like there are additional tests that needs to be skipped. however I can't be sure. See: 
https://app.circleci.com/pipelines/github/pytorch/pytorch/322056/workflows/fe2c23e9-fe0a-4d1f-8b55-2877c77c48eb/jobs/13459579

IvanYashchuk committed:
Update on "Update internal code for torch.linalg.solve"


Replace linalg_solve_helper with `lu_stub` + `lu_solve_stub`.
Once `lu_stub` and `lu_solve_stub` have cuSOLVER-based codepath,
`torch.linalg.solve` will have it as well.

[ghstack-poisoned]

mruberry commented:
This entire stack looks good to me, @nikitaved would you take a closer look just to sanity check?

IvanYashchuk commented:
I need to investigate the Windows build failure.

IvanYashchuk committed:
Update on "Update internal code for torch.linalg.solve"


Replace linalg_solve_helper with `lu_stub` + `lu_solve_stub`.
Once `lu_stub` and `lu_solve_stub` have cuSOLVER-based codepath,
`torch.linalg.solve` will have it as well.

[ghstack-poisoned]

IvanYashchuk committed:
Update on "Update internal code for torch.linalg.solve"


Replace linalg_solve_helper with `lu_stub` + `lu_solve_stub`.
Once `lu_stub` and `lu_solve_stub` have cuSOLVER-based codepath,
`torch.linalg.solve` will have it as well.

[ghstack-poisoned]

nikitaved reviewed (commented):
Apologies for the delay! LGTM!

IvanYashchuk committed:
Update on "Update internal code for torch.linalg.solve"


Replace linalg_solve_helper with `lu_stub` + `lu_solve_stub`.
Once `lu_stub` and `lu_solve_stub` have cuSOLVER-based codepath,
`torch.linalg.solve` will have it as well.

[ghstack-poisoned]

mruberry reviewed (approved):
Stamped! Thanks @IvanYashchuk and @nikitaved!

mruberry commented:
@mruberry has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28248766).

IvanYashchuk commented:
Note: tests pass with `CUDA_LAUNCH_BLOCKING=1` in the above failing CI builds.

IvanYashchuk committed:
Update on "Update internal code for torch.linalg.solve"


Replace linalg_solve_helper with `lu_stub` + `lu_solve_stub`.
Once `lu_stub` and `lu_solve_stub` have cuSOLVER-based codepath,
`torch.linalg.solve` will have it as well.

Differential Revision: [D28248766](https://our.internmc.facebook.com/intern/diff/D28248766)

[ghstack-poisoned]

IvanYashchuk committed:
Update on "Update internal code for torch.linalg.solve"


Replace linalg_solve_helper with `lu_stub` + `lu_solve_stub`.
Once `lu_stub` and `lu_solve_stub` have cuSOLVER-based codepath,
`torch.linalg.solve` will have it as well.

Differential Revision: [D28248766](https://our.internmc.facebook.com/intern/diff/D28248766)

[ghstack-poisoned]

IvanYashchuk commented:
@mruberry, I've submitted the changes from the `ci-all` branch, here is the draft PR https://github.com/pytorch/pytorch/pull/57993 for the CI status; it seems that the problem with Windows builds is gone.

The solution was to add a barrier after the call to the MAGMA routine for the batched LU decomposition that prevents premature usage of the data in the next MAGMA calls.

mruberry commented:
@mruberry has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28379394).

walterddr reviewed (commented):
reverting this PR as it broke no-arch test on master.

IvanYashchuk committed:
Update on "Update internal code for torch.linalg.solve"


Replace linalg_solve_helper with `lu_stub` + `lu_solve_stub`.
Once `lu_stub` and `lu_solve_stub` have cuSOLVER-based codepath,
`torch.linalg.solve` will have it as well.

Differential Revision: [D28379394](https://our.internmc.facebook.com/intern/diff/D28379394)

[ghstack-poisoned]

IvanYashchuk commented:
@mruberry, the problem should be fixed now. The issue was that I forgot to update the dispatch key of `linalg_solve`. Since the implementation of this function now uses DispatchStub, we need to have `CPU, CUDA` dispatch key instead of `CompositeExplicitAutograd`.

walterddr commented:
not exactly sure what went wrong in previous round of ci-all tests in #57993 but it did show noarch tests passing 
(in fact those failed master tests were skipped on your ci-all PR)
See:
https://app.circleci.com/pipelines/github/pytorch/pytorch/322056/workflows/fe2c23e9-fe0a-4d1f-8b55-2877c77c48eb/jobs/13459579
vs.
https://app.circleci.com/pipelines/github/pytorch/pytorch/319696/workflows/34bd1c85-0d8d-4db5-b0f0-056b9fba17e8/jobs/13340149

I will try to dig deeper into the reason behind this

IvanYashchuk commented:
@walterddr,  `at::tensorsolve` started to use `at::linalg_solve` that is being modified in this PR instead of the deprecated `at::solve`. The PR for `at::tensorsolve` was merged 2 days ago https://github.com/pytorch/pytorch/pull/57741.
So it's just two related PRs merged simultaneously without being tested against each other. Now it should be safe to merge this PR, ci-all tests in #57993 pass except one failure that seems to be unrelated.

mruberry commented:
@mruberry has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28627408).

mruberry commented:
@IvanYashchuk Sorry I missed the update on this. I've reimported it so we can land it. Just ping me if this is still good to go!

IvanYashchuk committed:
Update on "Update internal code for torch.linalg.solve"


Replace linalg_solve_helper with `lu_stub` + `lu_solve_stub`.
Once `lu_stub` and `lu_solve_stub` have cuSOLVER-based codepath,
`torch.linalg.solve` will have it as well.

Differential Revision: [D28627408](https://our.internmc.facebook.com/intern/diff/D28627408)

[ghstack-poisoned]

IvanYashchuk commented:
@mruberry, could you please import this again? I rebased this PR, let's get it merged.

The draft PR #57993 is for the CI-all status.

mruberry commented:
@mruberry has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28627408).
